sdks/go/pkg/beam/core/metrics/metrics.go

```diff
@@ -482,7 +482,39 @@ func (mr Results) AllMetrics() QueryResults {
 	return QueryResults{mr.counters, mr.distributions, mr.gauges}
 }
 
-// TODO(BEAM-11217): Implement Query(Filter) and metrics filtering
+// SingleResult interface facilitates metrics query filtering methods.
+type SingleResult interface {
+	Step() string
+	Name() string
+	Namespace() string
+}
+
+// Query allows metrics querying with filter. The filter takes the form of predicate function.
+// Ex: qr = pr.Metrics().Query(func(sr metrics.SingleResult) bool {
+//			return sr.Namespace() == test.namespace
+// })
```

Comment:
Nit: To format code blocks correctly in Godocs, you need to have them indented.
```suggestion
// Query allows metrics querying with filter. The filter takes the form of predicate function. Example:
//   qr = pr.Metrics().Query(func(sr metrics.SingleResult) bool {
//       return sr.Namespace() == test.namespace
//   })
```